### PR TITLE
fix some multithreading issues in Resource Notification

### PR DESCRIPTION
**Pull request by jukzi**
* adding missing final, volatile modifiers
* use of AtomicLong for IDs
* ResourceChangeListenerList uses CopyOnWriteArrayList

The javadoc of ResourceChangeListenerList.getListeners() was wrong:
subsequent adds of same listener (with different modifier) would have
changed the inside of the array.

### CODE REVIEW

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java [hunk 1]
```diff
@@ -79,29 +73,23 @@ public String toString() {
 	 * @param mask event types
 	 */
 	public synchronized void add(IResourceChangeListener listener, int mask) {
-		Assert.isNotNull(listener);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:82]: Why is this removed now?

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java [hunk 2]
```diff
@@ -79,29 +73,23 @@ public String toString() {
...
 			return;
 		}
 		ResourceChangeListenerList.ListenerEntry entry = new ResourceChangeListenerList.ListenerEntry(listener, mask);
-		final int oldSize = listeners.length;
+		final int oldSize = listeners.size();
 		// check for duplicates using identity
 		for (int i = 0; i < oldSize; ++i) {
-			if (listeners[i].listener == listener) {
-				removing(listeners[i].eventMask);
+			if (listeners.get(i).listener == listener) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:84]: `listeners.get(i)` could be extracted

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java [hunk 3]
```diff
@@ -79,29 +73,23 @@ public String toString() {
...
+				removing(listeners.get(i).eventMask);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:85]: This is interesting (unrelated to the current change). Javadoc says no effect, but here if the listener was added previously to listen with different mask, the mask will be changed.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java (hunk 1 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:82]: listener is final now

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java [hunk 4]
```diff
@@ -161,26 +141,26 @@ public boolean hasListenerFor(int event) {
 	 * @param listener the listener to remove
 	 */
 	public synchronized void remove(IResourceChangeListener listener) {
-		Assert.isNotNull(listener);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:164]: Why this is removed?

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java (hunk 2 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:84]: premature optimisation

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java (hunk 1 above)
Comment (iloveeclipse) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:82]: Sorry, this can't be true - it is argument to the public method.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/MarkerManager.java [hunk 5]
```diff
@@ -35,12 +36,12 @@ public class MarkerManager implements IManager {
 	private static final MarkerInfo[] NO_MARKER_INFO = new MarkerInfo[0];
 	private static final IMarker[] NO_MARKERS = new IMarker[0];
 	protected MarkerTypeDefinitionCache cache = new MarkerTypeDefinitionCache();
-	private long changeId = 0;
-	protected Map<IPath, MarkerSet> currentDeltas = null;
+	private AtomicLong changeId = new AtomicLong();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/MarkerManager.java:39]: final

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java [hunk 6]
```diff
@@ -118,8 +119,8 @@ public class Workspace extends PlatformObject implements IWorkspace, ICoreConsta
 	protected IMoveDeleteHook moveDeleteHook = null;
 	protected NatureManager natureManager;
 	protected FilterTypeManager filterManager;
-	protected long nextMarkerId = 0;
-	protected long nextNodeId = 1;
+	protected AtomicLong nextMarkerId = new AtomicLong();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java:122]: final, here and below

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java (hunk 1 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/ResourceChangeListenerList.java:82]: oh my bad

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/MarkerManager.java (hunk 5 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/MarkerManager.java:39]: good point.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java (hunk 6 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java:122]: thanks for review.
